service: killmode=cgroup|mixed, SendSIGKILL=no services are not multiprocess

grooverdan opened the pull request:
Refuse to start a service of this type as these are often database services that don't like having multiple instances trampling over their files.

closes #8630

test:
![image](https://user-images.githubusercontent.com/462287/51286711-44750d00-1a48-11e9-935a-ce2b1588e50b.png)

bl33pbl0p commented:
What if I remove SendSIGKILL=no, daemon-reload, and then try to start it?

grooverdan commented:
Then it will start a second instance. Is this a problem? systemd doesn't maintain service state after the timeout so the existence of a process in the cgroup and the current state of the service file are all that can be considered.

poettering commented on src/core/service.c:
```diff
@@ -1426,6 +1426,26 @@ static int service_spawn(
+                _cleanup_free_ char *f = NULL;
+                r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, UNIT(s)->cgroup_path, "cgroup.procs", &f);
+                if (r < 0)
+                        return r;
```
please use cg_is_empty_recursive() for this check

grooverdan replied:
I ended up getting `unit_warn_leftover_processes` to identify the process existing rather than separate call to `cg_is_empty_recursive`. This didn't affect the other calls to `unit_warn_leftover_processes`.

poettering commented on src/core/service.c:
```diff
@@ -1426,6 +1426,26 @@ static int service_spawn(
+                        /* text not empty, process must exist */
+                        if (*text != '\0') {
+                                r = -EINVAL;
+                                log_unit_error_errno(UNIT(s), r, "Will not start SendSIGKILL=no service of type KillMode=cgroup or mixed while process %s is still active", text);
```
hmm, could you please add a comment here, explaining why the combination of these settings means it's probably safe to refuse starting in this case.

poettering commented:
looks pretty ok, but please use the right helper call for this. That should simplify the patch quite a bit, and is a lot faster, in particular in cgroupsv2

poettering commented:
hmm, actually, the check is not at the right place: service_spawn() is used for any process we fork off for a service, including the ExecStop= stuff, where we don#t want this.

poettering commented:
this check should be placed where we currently invoke unit_warn_leftover_processes() already, i.e. in service_enter_start() and service_enter_start_pre()...

grooverdan commented:
Thanks for the prompt review. Rewritten to address these comments and used the existing `unit_warn_leftover_processes` function to perform the process check.

poettering commented on src/core/service.c:
```diff
@@ -2017,7 +2037,9 @@ static void service_enter_start(Service *s) {
-        unit_warn_leftover_processes(UNIT(s));
+        if (unit_warn_leftover_processes(UNIT(s)))
+                if (service_adverse_to_leftover_processes(s))
+                        goto fail;
```
You need to initialize `r` somehow if you jump to `fail`...

poettering commented on src/core/service.c:
```diff
@@ -2110,7 +2132,9 @@ static void service_enter_start_pre(Service *s) {
-                unit_warn_leftover_processes(UNIT(s));
+                if (unit_warn_leftover_processes(UNIT(s)))
+                        if (service_adverse_to_leftover_processes(s))
+                                goto fail;
```
similar here

poettering replied:
actually, the code in both occasions your new function is called is pretty much the same, hence I figure the invocation of `unit_warn_leftover_processes()` should maybe just move into `service_adverse_to_leftover_process()`? That'd be shorter then, and I like short.

poettering commented on src/core/service.c:
```diff
@@ -2006,6 +2006,26 @@ static void service_kill_control_process(Service *s) {
         }
 }
 
+static bool service_adverse_to_leftover_processes(Service *s) {
```
given that when you call this you need to initialize `r` to something maybe just return an `int` here, and then return an error like `-EBUSY` when we refuse to proceed like this, and 0 otherwise?

poettering commented on src/core/unit.c:
```diff
@@ -5396,27 +5401,35 @@ int unit_prepare_exec(Unit *u) {
+        (void) cg_kill_recursive(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, 0, 0, NULL, log_leftover, &ur);
 
-        (void) cg_kill_recursive(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, 0, 0, NULL, log_leftover, u);
+        return ur.result;
```
Hmm, I think it would be nicer if the logging callback that cg_kill_recursive() takes would be slightly altered, so that it returns a boolean. And then cg_kill_recursive() should return > 0 when at least one invocation of that function returns true, and 0 otherwise. Right now cg_kill_recursive() already returns > 0 when at least one process is killed, and letting the log function decide whether to return that would be only a minor extension the way I see it, and a natural one.

With that in place you don't need the small structure to pass data between this function and the log function, which would be nicer I think.

grooverdan commented:
updated. Comes out quite nicely. Even `systemctl start sleep` fails immediately.

![image](https://user-images.githubusercontent.com/462287/51449006-34c63300-1d7e-11e9-80b8-b152ddc66072.png)

poettering commented on src/core/unit.c:
```diff
@@ -4428,21 +4428,23 @@ int unit_make_transient(Unit *u) {
                         strna(comm),
                         signal_to_string(sig));
+
+        return false;
```
not that it matters at all, but in order to keep this similar to log_leftover() i think this should return true here

poettering commented on src/core/unit.c:
```diff
@@ -5394,29 +5396,31 @@ int unit_prepare_exec(Unit *u) {
+                return false;
 
-        log_unit_warning(userdata,
+        log_unit_warning((Unit *) userdata,
```
the explicit cast to Unit* is unnecessary, as in C "void*" casts implicitly to any type

poettering commented on src/core/unit.c:
```diff
@@ -5394,29 +5396,31 @@ int unit_prepare_exec(Unit *u) {
+                return false;
 
-        (void) cg_kill_recursive(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, 0, 0, NULL, log_leftover, u);
+        return cg_kill_recursive(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, 0, 0, NULL, log_leftover, u) == 1;
```
i think we should propagate the return as is (i.e. change the return type of unit_warn_leftover_processes() to "int"), and let the caller eat up any errors. It's more in line with the vast majority of other functions we have, and I think more generic, if we propagate errors always and eat them up as far up the call chain as appropriate but not needlessly far down.

i.e. make this function return "int", so that it returns < 0 on error, == 0 if no offending processes were found and > 0 if there were.

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -267,7 +267,8 @@ int cg_kill(
                         if (log_kill)
-                                log_kill(pid, sig, userdata);
+                                if (log_kill(pid, sig, userdata))
+                                        return 1;
```
a few lines further down there's currently this logic:

```c
if (ret == 0)
        ret = 1
```

i.e. this sets the return value to > 0 as soon as at least one process was killed. I think this should be conditionalized depending on whether log_kill is passed. i.e. the return value is determine by log_kill if it is set, and otherwise the return value is > 0 if at least one process was killed, as before. i.e. change that code to a simple:

```c
if (!log_kill && ret == 0)
        ret = 1;
```

poettering commented on src/core/service.c:
```diff
@@ -2006,6 +2006,27 @@ static void service_kill_control_process(Service *s) {
+            IN_SET(s->kill_context.kill_mode, KILL_MIXED, KILL_CONTROL_GROUP) &&
+            !s->kill_context.send_sigkill) {
+               log_unit_error(UNIT(s), "Will not start SendSIGKILL=no service of type KillMode=cgroup or mixed while processes exist");
+               return -EBUSY;
```
we now have a new syntax in our codebase to allow merging these last two lines in one, please use it:

```c
return log_unit_error_errno(UNIT(s), SYNTHETIC_ERRNO(EBUSY), "Will not start…");
```

poettering commented:
Looks pretty good, but a few more notes, see above. We are getting close.

I'd like to delay this to v242 however (i.e. not put this in the upcoming v241), since this has a chance of breaking some users and v241 is supposed to be a bugfix only release. Hope that makes sense.

grooverdan commented:
Updated and rebased. At some point it was the case that 'systemctl start {service}' where there was a process existing terminated immediately. Currently it seems to wait. That would be a desirable outcome but I've lost what case achieved this. Suggestions/Feedback welcome.

bl33pbl0p commented:
Thanks, I have tested your patch, it works nicely now, and doesn't seem to wait in the start job anymore (which I think is the correct behaviour in this case).

One improvement would be the description, it should read " ... KillMode=control-group ..." instead of cgroup (there is no cgroup option) which could confuse people.

Just curious though (about the usecase), why is not bumping the TimeoutStopSec= to infinity a possible option here (or better, JobRunningTimeoutSec= (we could add one for stop too)), that would mean your unit stays in "deactivating" state as long as everything doesn't clean up, and you effectively keep start jobs waiting until that is done, getting the semantics you want here, without denying requests outright, or failing them.

grooverdan commented:
Thanks for testing. Glad you didn't see a wait in the starting case.

Changed KillMode=control-group as requested.

Setting absolute timeouts on database services is quite hard as there is a very variable amount of shutdown/startup time. I added `EXTEND_TIMEOUT_USEC=` as an `sd_notify` message back in v236 for such a case. Evidently the implemented case in MariaDB (at the time #8630 was written) didn't send enough messages at the right time.

As getting the right number of `EXTEND_TIMEOUT_USEC=` messages going at the right time is a more difficult problem. This PR adds a sufficient safety mechanism to not add more processes while other timeout around processes are terminating. Without this PR, processes from two different services could running, of which a sysadmin or OOMKiller could accidentally kill the wrong one(s).

grooverdan commented:
Thanks @bl33pbl0p and @poettering
